pkg/applications/installer.go

```diff
@@ -36,57 +37,73 @@ import (
 	if err != nil {
-		return fmt.Errorf("failed to initialize source provider: %w", err)
+		return nil, fmt.Errorf("failed to initialize source provider: %w", err)
 	}
 
 	downloadDest, err := os.MkdirTemp(a.ApplicationCache, applicationInstallation.Namespace+"-"+applicationInstallation.Name)
 	if err != nil {
-		return fmt.Errorf("failed to create temporary directory where application source will be downloaded: %w", err)
+		return nil, fmt.Errorf("failed to create temporary directory where application source will be downloaded: %w", err)
 	}
 	defer func() {
 		if err := os.RemoveAll(downloadDest); err != nil {
 			log.Error("failed to remove temporary directory where application source has been downloaded: %s", err)
 		}
 	}()
 
+	templateProvider, err := providers.NewTemplateProvider(ctx, a.Kubeconfig, a.ApplicationCache, log, applicationInstallation, applicationInstallation.Status.ApplicationVersion.Template.Method)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize template provider: %w", err)
+	}
+
 	// start reconciliation
 	if err := a.reconcileNamespace(ctx, log, applicationInstallation, userClient); err != nil {
-		return err
+		return nil, err
 	}
 
 	appSourcePath, err := sourceProvider.DownloadSource(downloadDest)
 	if err != nil {
-		return fmt.Errorf("failed to download application source: %w", err)
+		return nil, fmt.Errorf("failed to download application source: %w", err)
 	}
 
-	log.Debugw("application successfully downloaded", "location", appSourcePath)
-	return nil
+	return templateProvider.InstallOrUpgrade(appSourcePath, applicationInstallation)
 }
 
 // Delete uninstalls the application and deletes the namespace where the application was installed if necessary.
-func (a *ApplicationManager) Delete(ctx context.Context, log *zap.SugaredLogger, seedClient ctrlruntimeclient.Client, userClient ctrlruntimeclient.Client, applicationInstallation *appskubermaticv1.ApplicationInstallation) error {
-	// todo logic to uninstall application
-	return a.deleteNamespace(ctx, log, applicationInstallation, userClient)
+func (a *ApplicationManager) Delete(ctx context.Context, log *zap.SugaredLogger, seedClient ctrlruntimeclient.Client, userClient ctrlruntimeclient.Client, applicationInstallation *appskubermaticv1.ApplicationInstallation) (util.StatusUpdater, error) {
+	templateProvider, err := providers.NewTemplateProvider(ctx, a.Kubeconfig, a.ApplicationCache, log, applicationInstallation, applicationInstallation.Status.ApplicationVersion.Template.Method)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize template provider: %w", err)
+	}
+
+	statuUpdater, err := templateProvider.Uninstall(applicationInstallation)
+	if err != nil {
+		return nil, err
+	}
+
+	return statuUpdater, a.deleteNamespace(ctx, log, applicationInstallation, userClient)
```

Comment:
I prefer to rename the field to `managed` and keep the entire desired state in the CRD.  The finalizer a meant to wait for the deletion of an external resource before deleting the crd, allowing retries if the deletion of this external resource failed.

Reply:
I think renaming the field is not enough to prevent something like this, which I find very realistic for a beginner user
1. create an app in a namespace
2. install other things in their namespace
3. uninstall the app
4. accidentally destroy the other things in their namespace.

Reply:
fair enough